lite/dbprovider.go

```diff
@@ -0,0 +1,260 @@
+		} else {
+			// Found the latest full commit signed header.
+			shBz := itr.Value()
+			sh := types.SignedHeader{}
+			err := dbp.cdc.UnmarshalBinary(shBz, &sh)
+			if err != nil {
+				return FullCommit{}, err
+			} else {
+				lfc, err := dbp.fillFullCommit(sh)
+				if err == nil {
+					dbp.logger.Info("DBProvider.LatestFullCommit() found latest.", "height", lfc.Height())
+					return lfc, nil
+				} else {
+					dbp.logger.Info("DBProvider.LatestFullCommit() got error", "lfc", lfc)
+					dbp.logger.Info(fmt.Sprintf("%+v", err))
+					return lfc, err
+				}
+			}
+		}
+	}
+	return FullCommit{}, lerr.ErrCommitNotFound()
+}
+
+func (dbp *DBProvider) ValidatorSet(chainID string, height int64) (valset *types.ValidatorSet, err error) {
+	return dbp.getValidatorSet(chainID, height)
+}
+
+func (dbp *DBProvider) getValidatorSet(chainID string, height int64) (valset *types.ValidatorSet, err error) {
+	vsBz := dbp.db.Get(validatorSetKey(chainID, height))
+	if vsBz == nil {
+		err = lerr.ErrMissingValidators(chainID, height)
+		return
+	}
+	err = dbp.cdc.UnmarshalBinary(vsBz, &valset)
+	if err != nil {
+		return
+	}
+	valset.TotalVotingPower() // to test deep equality.
+	return
+}
+
+func (dbp *DBProvider) fillFullCommit(sh types.SignedHeader) (FullCommit, error) {
+	var chainID = sh.ChainID
+	var height = sh.Height
+	var valset, nextValset *types.ValidatorSet
+	// Load the validator set.
+	valset, err := dbp.getValidatorSet(chainID, height)
+	if err != nil {
+		return FullCommit{}, err
+	}
+	// Load the next validator set.
+	nextValset, err = dbp.getValidatorSet(chainID, height+1)
+	if err != nil {
+		return FullCommit{}, err
+	}
+	// Return filled FullCommit.
+	return FullCommit{
+		SignedHeader:   sh,
+		Validators:     valset,
+		NextValidators: nextValset,
+	}, nil
+}
+
+func (dbp *DBProvider) deleteAfterN(chainID string, after int) error {
+
+	dbp.logger.Info("DBProvider.deleteAfterN()...", "chainID", chainID, "after", after)
+
+	itr := dbp.db.ReverseIterator(
+		signedHeaderKey(chainID, 1<<63-1),
+		signedHeaderKey(chainID, 0),
+	)
+	defer itr.Close()
+
+	var lastHeight int64 = 1<<63 - 1
+	var numSeen = 0
+	var numDeleted = 0
+
+	for itr.Valid() {
+		key := itr.Key()
+		_, height, ok := parseChainKeyPrefix(key)
+		if !ok {
+			return fmt.Errorf("unexpected key %v", key)
+		} else {
+			if height < lastHeight {
+				lastHeight = height
+				numSeen += 1
+			}
+			if numSeen > after {
+				dbp.db.Delete(key)
+				numDeleted += 1
+			}
+		}
+		itr.Next()
+	}
+
+	dbp.logger.Info(fmt.Sprintf("DBProvider.deleteAfterN() deleted %v items\n", numDeleted))
+	return nil
+}
+
+//----------------------------------------
+
+func signedHeaderKey(chainID string, height int64) []byte {
+	return []byte(fmt.Sprintf("%s/%010d/sh", chainID, height))
+}
+
+var signedHeaderKeyPattern = regexp.MustCompile(`([^/]+)/([0-9]*)/sh`)
+
+func parseSignedHeaderKey(key []byte) (chainID string, height int64, ok bool) {
+	submatch := signedHeaderKeyPattern.FindSubmatch(key)
```

Comment:
parseSignedHeaderKey and parseChainKeyPrefix funcs are the same except the key pattern. we can extract the common code into a separate function